scheds/rust/scx_bpfland/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,490 @@
+	 * architecture), relying on the built-in idle selection logic.
+	 */
+	if (!builtin_idle)
+		return prev_cpu;
+
+	/*
+	 * Find another idle CPU and potentially migrate the task.
+	 */
+	cpu = scx_bpf_select_cpu_dfl(p, prev_cpu, wake_flags, &is_idle);
+	if (is_idle) {
+		dispatch_direct_cpu(p, cpu);
+		return cpu;
+	}
+
+	return prev_cpu;
+}
+
+void BPF_STRUCT_OPS(bpfland_enqueue, struct task_struct *p, u64 enq_flags)
+{
+	/*
+	 * Always dispatch per-CPU kthreads immediately.
+	 *
+	 * This allows to prioritize critical kernel threads that may
+	 * potentially slow down the entire system if they are blocked for too
+	 * long (i.e., ksoftirqd/N, rcuop/N, etc.).
+	 */
+	if (is_kthread(p) && p->nr_cpus_allowed == 1) {
+		dispatch_kthread(p, enq_flags);
+		return;
+	}
+
+	/*
+	 * Dispatch all the other tasks that were not dispatched directly in
+	 * select_cpu().
+	 */
+	dispatch_task(p, enq_flags);
+}
+
+void BPF_STRUCT_OPS(bpfland_dispatch, s32 cpu, struct task_struct *prev)
+{
+	/*
+	 * First consume directly dispatched tasks, so that they can
+	 * immediately use the CPU assigned in select_cpu().
+	 */
+	if (scx_bpf_consume(cpu_to_dsq(cpu)))
+		return;
+
+	/*
+	 * Then always consume interactive tasks before regular tasks.
+	 *
+	 * This is fine and we shouldn't have starvation, because interactive
+	 * tasks are classified by their amount of voluntary context switches,
+	 * so they should naturally release the CPU quickly and give a chance
+	 * to the regular tasks to run.
+	 */
+	if (scx_bpf_consume(PRIO_DSQ))
+		return;
+
+	/*
+	 * Lastly, consume regular tasks from the shared DSQ.
+	 */
+	scx_bpf_consume(SHARED_DSQ);
+}
+
+void BPF_STRUCT_OPS(bpfland_running, struct task_struct *p)
+{
+	/* Update global vruntime */
+	if (vtime_before(vtime_now, p->scx.dsq_vtime))
+		vtime_now = p->scx.dsq_vtime;
+
+	__sync_fetch_and_add(&nr_running, 1);
+}
+
+/*
+ * Update task statistics when the task is releasing the CPU (either
+ * voluntarily or because it expires its assigned time slice).
+ */
+void BPF_STRUCT_OPS(bpfland_stopping, struct task_struct *p, bool runnable)
+{
+	u64 slice = slice_ns - p->scx.slice;
+	u64 weight = p->scx.weight;
+	u64 now = bpf_ktime_get_ns();
+	struct task_ctx *tctx;
+
+	__sync_fetch_and_sub(&nr_running, 1);
+
+	tctx = lookup_task_ctx(p);
+	if (!tctx)
+		return;
+
+	/*
+	 * Classify interactive tasks based on the average amount of their
+	 * voluntary context switches and automatically boost their priority
+	 * (that will affect their vruntime).
+	 */
+	if (nvcsw_thresh && weight_boost) {
+		if (tctx->avg_nvcsw >= nvcsw_thresh)
+			tctx->is_interactive = true;
+		if (tctx->is_interactive)
+			weight *= weight_boost;
+	}
+
+	/*
+	 * Evaluate weighted time slice and assign the task a time slice within
+	 * the range [1, slice_ns]. Assigning a minimum slice of 1 ensures
+	 * strictly monotonic progress of the vruntime.
+	 */
+	slice = MAX(slice * 100 / weight, 1);
+	slice = MIN(slice, slice_ns);
```

Comment:
Hmm, so for vtime we theoretically _want_ some tasks to accumulate more than a slice worth of vtime if they have a low weight. Say they ran for `slice_ns` and had a weight of 1. That should mean that they should get 1/100 as much CPU as a task that with the default weight of 100. If we did `slice = MIN(slice, slice_ns)`, then they'll always accumulate the same as a task with weight 100.

Reply:
Right.. the idea here was to put an upper limit for the slice, to make sure it never becomes too big, but `slice_ns` is probably too small, because it would "smooth" too much the different tasks' priorities. Maybe I should introduce a `slice_ns_max` or limit the slice to `slice_ns` and then apply the weight? (I'll do some tests with both)